network/local.go

```diff
@@ -0,0 +1,459 @@
+	}
+	if ll.ctx.IsListening(addr) {
+		return fmt.Errorf("%s is already listening: can't listen again", addr)
+	}
+	ll.addr = addr
+	return nil
+}
+
+// Listen implements the Listener interface
+func (ll *LocalListener) Listen(fn func(Conn)) error {
+	ll.Lock()
+	ll.quit = make(chan bool)
+	ll.ctx.Listening(ll.addr, fn)
+	ll.listening = true
+	ll.Unlock()
+
+	<-ll.quit
+	return nil
+}
+
+// Stop implements the Listener interface
+func (ll *LocalListener) Stop() error {
+	ll.Lock()
+	defer ll.Unlock()
+	ll.ctx.StopListening(ll.addr)
+	if ll.listening {
+		close(ll.quit)
+	}
+	ll.listening = false
+	return nil
+}
+
+// Address returns the listening address used.
+func (ll *LocalListener) Address() Address {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.addr
+}
+
+// Listening returns true if this Listener is actually listening for any
+// incoming connections.
+func (ll *LocalListener) Listening() bool {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.listening
+}
+
+// LocalHost is a Host implementation using LocalConn and LocalListener as
+// the underlying means of communication. It is a implementation of the Host
+// interface.
+type LocalHost struct {
+	addr Address
+	*LocalListener
+	ctx *LocalContext
+}
+
+// NewLocalHost returns a fresh Host using Local communication that will listen
+// on the given addr.
+func NewLocalHost(addr Address) (*LocalHost, error) {
+	return NewLocalHostWithContext(defaultLocalContext, addr)
+}
+
+// NewLocalHostWithContext is similar to NewLocalHost but specify which
+// LocalContext to use for communicating.
+func NewLocalHostWithContext(ctx *LocalContext, addr Address) (*LocalHost, error) {
+	lh := &LocalHost{
+		addr: addr,
+		ctx:  ctx,
+	}
+	var err error
+	lh.LocalListener, err = NewLocalListenerWithContext(ctx, addr)
+	return lh, err
+
+}
+
+// Connect implements the Host interface.
+func (lh *LocalHost) Connect(addr Address) (Conn, error) {
+	if addr.ConnType() != Local {
+		return nil, errors.New("Can't connect to non-Local address")
+	}
+	var finalErr error
+	for i := 0; i < MaxRetryConnect; i++ {
+		c, err := NewLocalConnWithContext(lh.ctx, lh.addr, addr)
+		if err == nil {
+			return c, nil
+		}
+		finalErr = err
+		time.Sleep(WaitRetry)
+	}
+	return nil, finalErr
+
+}
+
+// NewLocalClient returns Client that uses the Local communication layer.
+func NewLocalClient() *Client {
+	return NewLocalClientWithContext(defaultLocalContext)
+}
+
+// NewLocalClientWithContext is similar to NewLocalClient but takes a specific
+// LocalContext to communicate.
+func NewLocalClientWithContext(ctx *LocalContext) *Client {
+	fn := func(own, remote *ServerIdentity) (Conn, error) {
+		return NewLocalConnWithContext(ctx, own.Address, remote.Address)
+	}
+	return newClient(fn)
+
+}
+
+// NewLocalAddress returns an Address of type Local with the given raw addr.
+func NewLocalAddress(addr string) Address {
+	return NewAddress(Local, addr)
+}
+
+/*// GetStatus implements the Host interface*/
```

Comment:
That's actually left for a reason I forgot to mention... :)
What to do with regards to Status? I can't make them implement the interface because the interface uses something sda specific (`type Status map[string]string` is in sda package). So for the moment I dropped them but now we don't have access to the statistics of the network layer (which is not great). Any plans / solutions ?
On top of my head, I would remove the specific `Status` type and then let `Router` / `Host` / `Conn` / `Listener` implements `GetStatus() map[string]string`.